src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,254 @@
+--
+--   Computes @let@-bindings from the given variables, composes the resulting
+--   substitutions and applies the substitution on the expression.
+buildLet :: IR.Expr -> [IR.VarName] -> Converter IR.Expr
+buildLet expr []   = return expr
+buildLet expr vars = do
+  let srcSpan = IR.exprSrcSpan expr
+  (binds, substs) <- buildBinds srcSpan vars
+  return (IR.Let srcSpan binds (applySubst (composeSubsts substs) expr)
+          (IR.exprTypeScheme expr))
+
+-- | Converts the list containing variables into @let@-bindings where
+--   the variable pattern is a fresh variable and the right-hand side is
+--   a variable that occurred multiple times on right hand sides.
+--
+--   Also computes substitutions mapping given variables to fresh variables.
+--   Returns the generated let-bindings and the substitutions.
+buildBinds :: SrcSpan -> [IR.VarName] -> Converter ([IR.Bind], [Subst IR.Expr])
+buildBinds srcSpan = mapAndUnzipM buildBind
+ where
+  buildBind :: IR.VarName -> Converter (IR.Bind, Subst IR.Expr)
+  buildBind varName = do
+    varName' <- freshHaskellQName varName
+    let subst            = singleSubst' varName
+          (\s -> IR.Var s varName' Nothing)
+        -- This type scheme has wrong source span information.
+        rhs              = IR.Var srcSpan varName Nothing
+        Just varPatIdent = IR.identFromQName varName'
+        varPat           = IR.VarPat srcSpan varPatIdent Nothing False
+        bind             = IR.Bind srcSpan varPat rhs
+    return (bind, subst)
+
+-- | Counts all variable names on right-hand sides of expression.
+--   Shadowed variables and variables from the list are not counted.
+--   Variables introduced on the left side of a @case@-alternative and @let@
+--   expressions are not counted.
+countVarNamesExcept :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
+countVarNamesExcept = countVarNamesWith (++) elem
+
+-- | Counts all variable names on right-hand sides of expression that are also
+--   contained by the given list.
+--   Shadowed variables and variables from the list are not counted.
+--   Variables introduced on the left side of a @case@-alternative and @let@
+--   expressions are not counted.
+countVarNamesOnly :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
+countVarNamesOnly = countVarNamesWith const notElem
+
+-- | A generalized recursive counting for variables in expressions.
+--   The first arguments is the function that is used to combine
+--   the given list of variable names with variable names from left hand sides
+--   of @case@-alternatives or lambda abstractions.
+--   The second argument is the a predicate that determines if a variable
+--   is added to the map.
+countVarNamesWith :: ([IR.VarName] -> [IR.VarName] -> [IR.VarName])
+                  -> (IR.VarName -> [IR.VarName] -> Bool)
+                  -> [IR.VarName]
+                  -> IR.Expr
+                  -> Map IR.VarName Integer
```

Comment:
You overengineered this function a little bit which makes it very hard to understand. The definition gets much easier without predicates and accumulator. In order to implement shadowing you just have to remove entries from the map, for example.

```haskell
countVarNames :: IR.Expr -> Map IR.VarName Integer
countVarNames (IR.Lambda _ args rhs _) 
  = countVarNames rhs `Map.withoutKeys` Set.fromList (map IR.varPatQName)
```

The intended behavior of `countVarNamesOnly` can then be obtained by restricting the key set of the map.

```haskell
countVarNamesOnly :: [IR.VarName] -> IR.Expr -> Map IR.VarName Integer
countVarNamesOnly varNames expr
  = countVarNames expr `Map.restirctKeys` Set.fromList varName 
```

This fixes the bug pointed out above automatically.

Reply:
That is a really good idea. I did not think of that. 
My intention was to avoid code duplication but that made the function way to complicated.